### ISSUE TITLE
Show Feature Count shows wrong numbers with rule-based classification

### ISSUE THREAD

**Issue by pathmapper** [qgis/QGIS#40206]
**Describe the bug**

`Show Feature Count` shows wrong numbers with rule-based classification.

**How to Reproduce**

1. Open the following sample file with QGIS: [sample_layer_definition_file.zip](https://github.com/qgis/QGIS/files/5574343/sample_layer_definition_file.zip)
2. Open contextmenu of the layer and click on `Show Feature Count`.
3. See error: Feature Count for _Viersen_, _Krefeld_ and _Wesel_ is wrong, e.g. click on "Test" in the "Edit Rule" window, compare with attribute table or visually count the features on the canvas. 

<img src="https://user-images.githubusercontent.com/20856381/99816068-7aefa780-2b4b-11eb-86d1-0b3a44596f0e.png" width="300">

![grafik](https://user-images.githubusercontent.com/20856381/99815858-2f3cfe00-2b4b-11eb-8990-f87a3ddf910d.png)

**QGIS and OS versions**

QGIS version | 3.16.0-Hannover | QGIS code revision | 43b64b13f3
-- | -- | -- | --
Compiled against Qt | 5.11.3 | Running against Qt | 5.11.3
Compiled against GDAL/OGR | 2.4.0 | Running against GDAL/OGR | 2.4.0
Compiled against GEOS | 3.7.1-CAPI-1.11.1 | Running against GEOS | 3.7.1-CAPI-1.11.1 27a5e771
Compiled against SQLite | 3.27.2 | Running against SQLite | 3.27.2
PostgreSQL Client Version | 11.9 (Debian 11.9-0+deb10u1) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.10.4
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | Debian GNU/Linux 10 (buster)
Active python plugins | ClipMultipleLayers; quick_map_services; scriptrunner3; LAStools; canvas_clipper; savedialog; loadthemall; clipper; kmltools; qgis-maptiler-plugin; MapsPrinter; Discovery; mmqgis; pluginbuilder3; vector_tiles_reader; layerGroupFilter; plugin_reloader; realcentroid; mask; MetaSearch; db_manager; processing

**Additional context**

The similiar issue https://github.com/qgis/QGIS/issues/39318 was closed because it was not reproducible.

There's something wrong in general with `Show feature count` for rule-based classification: https://github.com/qgis/QGIS/issues/27893, https://github.com/qgis/QGIS/issues/27892

**Comment by gioman** [comment]
@pathmapper same as https://github.com/qgis/QGIS/issues/37799 ?

**Comment by roya0045** [comment]
@gioman I'd say yes.

**Comment by pathmapper** [comment]
> same as #37799 ?

@gioman I'd say no, because if I understood it correct #37799 is not about showing wrong numbers, instead it's about showing the total number of features even if a filter is applied. I'd say that #37799 is a duplicate of this feature request: https://github.com/qgis/QGIS/issues/29458

The issue reported in this ticket here is about showing wrong numbers. The count number **23** shown is totally wrong and has nothing to do with the total number of features. **23** is the total number of features minus all features which are not `"KREIS" = 'Kleve'`.

Correct would be:
|Kreis|feature count|
|:---|---:|
|Viersen|9|
|Krefeld|1|
|Wesel|13|
|Kleve|16|
|**total**|**39**|

**Comment by roya0045** [comment]
Did you toggle the count when you had only two classes? This would indicate so.

**Comment by pathmapper** [comment]
> Did you toggle the count when you had only two classes? 

@roya0045 no, just reproduced the issue: 

![wrong_feature_count](https://user-images.githubusercontent.com/20856381/99838416-66220c80-2b69-11eb-9532-e12cd68e2907.gif)

**Comment by roya0045** [comment]
What if you use the category symbology first to get the classes and then switch to rule-based? Or used the refine selected rules button instead of doing it all one by one?

**Comment by pathmapper** [comment]
> What if you use the category symbology first to get the classes and then switch to rule-based?

In this case the numbers are correct:
![use_category_first](https://user-images.githubusercontent.com/20856381/99869861-99e34d80-2bce-11eb-866c-9393343168f9.gif)

> Or used the refine selected rules button instead of doing it all one by one?

If you mean something like the following, it works also: 

![refine_selected_rules](https://user-images.githubusercontent.com/20856381/99869954-6d7c0100-2bcf-11eb-976e-91bf43cee9bb.gif)

Here is another one using the `Add Rule` button:

![Use_plus_button](https://user-images.githubusercontent.com/20856381/99870002-dc595a00-2bcf-11eb-91d6-8467c13fd0d2.gif)

---

So it looks like the bug here which leads to wrong feature count numbers is regarding the use of `copy`/`paste` of rules via the contextmenu:

<img src="https://user-images.githubusercontent.com/20856381/99870148-170fc200-2bd1-11eb-8fec-32083f4285e9.png" width="300">

**Comment by roya0045** [comment]
Ok I think this is enough to isolate the issue.

When counting features the rulekeys for feature is used
https://github.com/qgis/QGIS/blob/213df8df2334b538a42b16b941b70fe1b2505c6e/src/core/qgsvectorlayerfeaturecounter.cpp#L75
In the rule based render the keys are generated at the start
https://github.com/qgis/QGIS/blob/842e765c49c4da9e0e28a0e3d596e7807c48bfad/src/core/symbology/qgsrulebasedrenderer.cpp#L52

But to maintain some usability those keys are not changed when cloning.
https://github.com/qgis/QGIS/blob/842e765c49c4da9e0e28a0e3d596e7807c48bfad/src/core/symbology/qgsrulebasedrenderer.cpp#L1039

@nyalldawson would it be a good idea to use the filter expression as a key? I know that this can be more complex with nested symbols as there can be duplicates in different branches. But my older PR has code to get a full comprehensive expression string for each 'node', this would solve that issue.

**Comment by beerockxs** [comment]
Using 3.22-1, I have this issue without having copy-pasted rules. I created a rule based symbology and refined it for two additional levels using the "Refine Selected Rules" button and adding further categories:

![grafik](https://user-images.githubusercontent.com/16764567/148215627-63d7ef7d-1c2b-4cab-a5a9-d8a916673c9e.png)

**Comment by pathmapper** [comment]
> But to maintain some usability those keys are not changed when cloning.

Yep, if the style is exported as `.qml`, the first three rules are having the same key.

![grafik](https://user-images.githubusercontent.com/20856381/189667289-f8244e70-3ed0-4eba-ac65-f2c5d174024e.png)

If you manually edit the `.qml` so that the keys are unique, the style works as expected.
